### PR TITLE
M2P-501 Create feature switch for customizable options support

### PR DESCRIPTION
**Pull request by dumega**
# Description
To enable/disable customizable options support by feature switch

Fixes: https://boltpay.atlassian.net/browse/M2P-501

#changelog Create feature switch for customizable options support

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: Helper/Cart.php [hunk 1]
```diff
@@ -1357,18 +1357,28 @@ function ($item) use ($imageHelper, &$totalAmount, &$diff, $storeId, $currencyCo
...
+                if ($this->deciderHelper->isCustomizableOptionsSupport()) {
+                    $itemProduct = $item->getProduct();
+                    $customizableOptions = $this->getProductCustomizableOptions($itemProduct);
+                    $itemSku = trim($item->getSku());
+                    
+                    if ($customizableOptions) {
+                        $itemSku = $this->getProductActualSkuByCustomizableOptions($itemSku, $customizableOptions);
+                    }
+                    
+                    $_product = $this->productRepository->get($itemSku);
```
Comment (vitaliyreznikov) [inline Helper/Cart.php:1370]: What if trying to catch error here and if so continue with `$itemProduct`?
What do you think about it?

Comment (vitaliyreznikov) [reply Helper/Cart.php:1370]: Another point, I think we shouldn't search product by sku if we didn't change it. We just added additional SQL calls without benefits.
So something like
```
                    if ($customizableOptions) {
                        $itemSku = $this->getProductActualSkuByCustomizableOptions($itemSku, $customizableOptions);
                         $_product = $this->productRepository->get($itemSku);
                    }
```

Comment (vitaliyreznikov) [reply Helper/Cart.php:1370]: Also, can we add productId to bugsnag record if we have any exceptions?

**Comment by dumega** [comment]
@vitaliyreznikov 

> Why do we get the product by SKU instead of ID?

Please refer to this PR https://github.com/BoltApp/bolt-magento2/pull/1089, cause from the quote item, we can not get the actual product directly if it is variable product.

And I have updated the PR as you suggest

File: Helper/Cart.php [hunk 2]
```diff
@@ -1360,18 +1360,36 @@ function ($item) use ($imageHelper, &$totalAmount, &$diff, $storeId, $currencyCo
...
+                                'product_id' => $item->getProductId(),
+                                'product_name' => $item->getName(),
+                                'product_sku' => $item->getSku()
+                            ]);
+                        });
+                        $this->bugsnag->notifyError('Could not retrieve product', $e->getMessage());
+                        $customizableOptions = null;
+                    }
+                }
+                if (empty($customizableOptions)) {
```
Comment (vitaliyreznikov) [inline Helper/Cart.php:1386]: I know this code works, but could you add `$customizableOptions = null;` before line 1365?

Comment (dumega) [reply Helper/Cart.php:1386]: Done

Comment (vitaliyreznikov) [reply Helper/Cart.php:1386]: Could you push your change?

**Comment by dumega** [comment]
@vitaliyreznikov The snippet within `if ($this->deciderHelper->isCustomizableOptionsSupport()) {` block is not only for customizable options but also for fixing the update-cart endpoint to edit qty of variable product. So whether the $customizableOptions is empty or not, we woud always use $_product = $this->productRepository->get($itemSku); to retrieve the product.

**Comment by vitaliyreznikov** [comment]
@dumega I am ok to merge this for now, but when you are online could you please explain why we need `$_product = $this->productRepository->get($itemSku); ` even if `$customizableOptions is empty`?

**Review by vitaliyreznikov — APPROVED** [review]

**Comment by dumega** [comment]
@vitaliyreznikov That's for get the product which is variable type. If the quote item is variable product, $item->getProductId(); just returns its parent product id, and we have to retrieve its proper id via sku.

**Comment by vitaliyreznikov** [comment]
Can we narrow down this assignment to `variable type` products only then?

**Review by danac-gs — APPROVED** [review]

File: Helper/Cart.php [hunk 3]
```diff
@@ -1360,18 +1360,53 @@ function ($item) use ($imageHelper, &$totalAmount, &$diff, $storeId, $currencyCo
...
+                $customizableOptions = null;
                 $itemSku = trim($item->getSku());
+                $itemReference = $item->getProductId();
+                $itemName = $item->getName();
 
-                if ($customizableOptions) {
-                    $itemSku = $this->getProductActualSkuByCustomizableOptions($itemSku, $customizableOptions);
+                if ($this->deciderHelper->isCustomizableOptionsSupport()) {
+                    try {
+
```
Comment (vitaliyreznikov) [inline Helper/Cart.php:1372]: nit: remove empty line

**Review by vitaliyreznikov — APPROVED** [review]

File: Helper/Cart.php (hunk 3 above)
Comment (dumega) [reply Helper/Cart.php]: Fixed

**Review by danac-gs — APPROVED** [review]

**Pull request merged**